androidApp/src/main/java/com/example/tasky/android/login/screen/LoginScreen.kt

```diff
@@ -1,23 +1,164 @@
 package com.example.tasky.android.login.screen
 
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.Button
+import androidx.compose.material3.MaterialTheme.typography
 import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.text.LinkAnnotation
+import androidx.compose.ui.text.SpanStyle
+import androidx.compose.ui.text.TextLinkStyles
+import androidx.compose.ui.text.buildAnnotatedString
+import androidx.compose.ui.text.withLink
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
 import androidx.navigation.NavGraphBuilder
 import androidx.navigation.compose.composable
-import com.example.tasky.Greeting
+import com.example.tasky.android.login.components.LoginPasswordTextField
+import com.example.tasky.android.login.components.LoginTextField
+import com.example.tasky.android.theme.Black
+import com.example.tasky.android.theme.Gray
+import com.example.tasky.android.theme.LightBlue
+import com.example.tasky.android.theme.MyApplicationTheme
 import kotlinx.serialization.Serializable
 
 fun NavGraphBuilder.loginScreen() {
     composable<Login> {
-        LoginScreen()
+        LoginScreen(
+            LoginScreenState(),
+            object : LoginScreenEvents {
+                override fun onClickToSignUp() {
+                }
+
+                override fun onClickLogin() {
+                }
+
+                override fun updateLoginScreenState(newState: LoginScreenState) {
+                }
+            },
```

Comment:
Instead of relying on these large, traditional callbacks, have you thought about using a lambda expression together with a sealed interface instead? That's a more modern and Kotlin idiomatic approach:
```
sealed interface LoginScreenEvent {
    data object OnClickToSignUp: LoginScreenEvent
    data object OnClickLogin: LoginScreenEvent
    data class OnEmailChange(val email: String): LoginScreenEvent
    ...
}
``` 
With that sealed interface, you can then have a function in your ViewModel to handle these:
```
class MyViewModel: ViewModel() {
    
    fun onEvent(event: LoginScreenEvent) {
        when(event) {
            is OnClickToSignUp -> ...
            is OnClickToLogin -> ...
            is OnEmailChange -> ...
        }
    }
}
```
And in the root screen, it just looks like this:
```
val state by viewModel.state.collectAsStateWithLifecycle()
LoginScreen(
    state = state,
    onEvent = viewModel::onEvent
)
```
This is the signature of`LoginScreen`:
```
@Composable
fun LoginScreen(
    state: LoginState,
    onEvent: (LoginScreenEvent) -> Unit
) {
    // ...
    
   TextField(
       value = state.email,
       onValueChange = {
           onEvent(LoginScreenEvent.OnEmailChange(it))
       },
       ...
   )
   // ...
}
```

Reply:
I am sorry that I do not understand the pros of this approach. Isn't it just changing a place to put the callbacks? 

Besides, sometimes these events may not be able to add in ViewModels, especially for navigation. As I know, the official docs do not recommend passing the navController to down to ui or ViewModels. Then, the only way I could think of would be adding an extra state in ViewModels and collect it in the navGraph. 

```
val state by viewModel.state.collectAsStateWithLifecycle()
val shouldNavigateToXXX by viewModel.navigateFlow.collectAsStateWithLifecycle()

LaunchedEffect(shouldNavigateToXXX) {
  if (shouldNavigateToXXX) {
    navController.navigateTo(XXX)
  }
}

LoginScreen(
    state = state,
    onEvent = viewModel::onEvent
)
```
Is there other way that could achieve this approach without adding these extra states